### PR TITLE
Add SLURM scripts for OLMo SFT with resume support

### PR DESCRIPTION
**Pull request by ferreirafabio**
Hey,

This PR adds SLURM-compatible scripts for running OLMo SFT training on clusters without Beaker/cloud infrastructure, as discussed with @hamishivi in #1325.

Moreover, tokenizing large SFT datasets can take >12 hours (even with over 300 cpu processes). Many SLURM clusters enforce preemptive scheduling and limited job time (e.g., 6 hours). Without resume support, any interruption means starting over from scratch. Therefore, this PR adds also checkpoint/resume support to `convert_sft_data_for_olmocore.py`, so users can simply resubmit the same script to continue where they left off.

## Quick overview: files added

```
  scripts/slurm/sft/
  ├── README.md                        # Documentation
  ├── prepare_dolci_think_data.sh      # Data prep for Think (~22B tokens)
  ├── prepare_dolci_instruct_data.sh   # Data prep for Instruct (~1.8B tokens)
  ├── train_dolci_think.sh             # Training script (lr=5e-5)
  └── train_dolci_instruct.sh          # Training script (lr=8e-5)
  open_instruct/
  └── test_checkpoint.py               # 15 unit tests for checkpoint functions

  scripts/data/
  └── convert_sft_data_for_olmocore.py # Modified: added resume support
```

Changes:

1. Resume checkpoint support for `convert_sft_data_for_olmocore.py`:
   - Add --resume flag to continue from last checkpoint after interruption
   - Add --checkpoint_interval to control checkpoint frequency (default 100k)
   - Checkpoints saved atomically to _checkpoint.json
   - Automatic cleanup on successful completion
   - Fixed shuffle seed (42) for reproducible resume

2. SLURM data preparation scripts:
   - `prepare_dolci_think_data.sh`: Tokenize Dolci-Think-SFT
   - `prepare_dolci_instruct_data.sh`: Tokenize Dolci-Instruct-SFT
   - Both include resume support for time-limited queues

3. SLURM training scripts (requires OLMo-core clone):
   - `train_dolci_think.sh`: Train Think SFT
   - `train_dolci_instruct.sh`: Train Instruct SFT
   - Hyperparameters from OLMo-3 paper (Table 47)
   - Uses `OLMOCORE_PATH` env var to locate OLMo-core installation

Usage:
  # Data preparation
  sbatch scripts/slurm/sft/prepare_dolci_think_data.sh

  # Training (after cloning OLMo-core)
  OLMOCORE_PATH=/path/to/OLMo-core \
  DATASET_PATH=./data/dolci_think_sft_tokenized \
  BASE_CKPT=/path/to/OLMo-3-7B \
  sbatch scripts/slurm/sft/train_dolci_think.sh

(Disclaimer: I created this PR with the help of Opus 4.5, e.g. for writing the scripts/slurm/sft/README).

I hope this can be helpful to all users / maintainers of open-instruct. Please let me know if there are questions / or requests for changes.

GPU_TESTS=[01KFGX8Q53EJTWGNKRFAN6M0N5](https://beaker.org/orgs/ai2/workspaces/open-instruct-dev/work/01KFGX8Q53EJTWGNKRFAN6M0N5?taskId=01KFGX8Q59403VFQ1NCZ7886M9&jobId=01KFGX8Q8Y73NPD74XMJTT1YDK)

### CODE REVIEW

File: scripts/data/convert_sft_data_for_olmocore.py [hunk 1]
```diff
@@ -213,29 +246,57 @@ def main(args: ConvertSFTDataArguments, tc: TokenizerConfig):
...
             file=sys.stdout,
             bar_format="{l_bar}{bar}{r_bar}\n",  # better printing in beaker
             mininterval=10.0,
+            initial=start_idx,  # Start progress bar from resume point
+            total=total_samples,
         )
     ):
+        # Skip already-processed samples when resuming
+        if idx < start_idx:
+            continue
```
[the remark this answers is not in the document]
Comment (ferreirafabio) [reply scripts/data/convert_sft_data_for_olmocore.py]: done

File: scripts/data/convert_sft_data_for_olmocore.py [hunk 2]
```diff
@@ -201,7 +233,8 @@ def main(args: ConvertSFTDataArguments, tc: TokenizerConfig):
         drop_dataset_source=False,
     )
 
-    train_dataset = train_dataset.shuffle()
+    # Use fixed seed for reproducible shuffling (required for resume)
+    train_dataset = train_dataset.shuffle(seed=42)
```
Comment (hamishivi) [inline scripts/data/convert_sft_data_for_olmocore.py]: Can we make this seed configurable, and default to 42?

**Comment by ferreirafabio** [comment]
@hamishivi thanks! Sure, quality checks should now pass and I made the seed configurable via --shuffle_seed (default: 42).

**Review by hamishivi — APPROVED** [review]
Looks good, could you add a changelog item?
I'm working on fixing/bypassing the gpu / unit tests, this is a problem in our CI, not your fault.

**Comment by ferreirafabio** [comment]
> Looks good, could you add a changelog item? I'm working on fixing/bypassing the gpu / unit tests, this is a problem in our CI, not your fault.

Done! Feel free to let me know if I can be of any help.

**Comment by finbarrtimbers** [comment]
I am pushing a PR to fix this; should be in later today. Thanks for your patience!

**Comment by finbarrtimbers** [comment]
Ok, I think we're good to go. Just waiting on the GPU tests to pass and then I'll merge it.

**Comment by ferreirafabio** [comment]
@hamishivi @finbarrtimbers not sure if anything is blocking on my end. I think we are good to merge and can close the PR? Thanks

**Comment by finbarrtimbers** [comment]
Yes! Trying now.
